runtime/jvmti/jvmtiCapability.c

```diff
@@ -540,55 +546,74 @@ mapCapabilitiesToEvents(J9JVMTIEnv * j9env, jvmtiCapabilities * capabilities, J9
 {
 	IDATA rc = 0;
 
-	if (capabilities->can_generate_single_step_events) {
-		rc |= eventHookFunction(j9env, JVMTI_EVENT_SINGLE_STEP);
-	}
-
-	if (capabilities->can_generate_breakpoint_events) {
-		rc |= eventHookFunction(j9env, JVMTI_EVENT_BREAKPOINT);
-	}
-
-	if (capabilities->can_generate_field_access_events) {
-		rc |= eventHookFunction(j9env, JVMTI_EVENT_FIELD_ACCESS);
-	}
-
-	if (capabilities->can_generate_field_modification_events) {
-		rc |= eventHookFunction(j9env, JVMTI_EVENT_FIELD_MODIFICATION);
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	J9JavaVM * vm = j9env->vm;
+	J9VMThread *mainThread = vm->mainThread;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	BOOLEAN skipHookReserve = vmFuncs->isCheckpointAllowed(mainThread)
+			&& vmFuncs->isDebugOnRestoreEnabled(mainThread);
+	/* Skip J9HookReserve for the events required by JDWP agent pre-checkpoint when DebugOnRestore is enabled,
+	 * these events will be registered post-restore if a JDWP agent is specified in the restore option file,
+	 * otherwise they are going to be unregistered by J9HookUnregister() which only clears J9HOOK_FLAG_HOOKED,
+	 * but not J9HOOK_FLAG_RESERVED.
+	 * J9HookUnreserve() might clear the flag set by other callers.
+	 */
+	if (!skipHookReserve)
+#endif /* defined(J9VM_OPT_CRIU_SUPPORT)*/
+	{
+		if (capabilities->can_generate_single_step_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_SINGLE_STEP);
+		}
+		if (capabilities->can_generate_breakpoint_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_BREAKPOINT);
+		}
+		if (capabilities->can_generate_field_access_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_FIELD_ACCESS);
+		}
+		if (capabilities->can_generate_field_modification_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_FIELD_MODIFICATION);
+		}
+		if (capabilities->can_generate_frame_pop_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_FRAME_POP);
+		}
+		if (capabilities->can_generate_monitor_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_MONITOR_CONTENDED_ENTER);
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_MONITOR_CONTENDED_ENTERED);
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_MONITOR_WAIT);
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_MONITOR_WAITED);
+		}
+#if JAVA_SPEC_VERSION >= 19
+		if (capabilities->can_support_virtual_threads) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_VIRTUAL_THREAD_START);
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_VIRTUAL_THREAD_END);
+		}
+#endif /* JAVA_SPEC_VERSION >= 19 */
+		if (capabilities->can_generate_garbage_collection_events) {
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_GARBAGE_COLLECTION_START);
+			rc |= eventHookFunction(j9env, JVMTI_EVENT_GARBAGE_COLLECTION_FINISH);
+		}
 	}
-
-	if (capabilities->can_generate_frame_pop_events) {
-		rc |= eventHookFunction(j9env, JVMTI_EVENT_FRAME_POP);
+	/* can't skip for criu */
```

Comment:
The `eventHookFunction()` is `reserveEvent` which invokes `J9HookReserve()` later https://github.com/eclipse-openj9/openj9/blob/c24c9542ffd415e5431ae2a2109129f6c55e4268/runtime/jvmti/jvmtiCapability.c#L414


Those capabilities have to be reserved if requested via `jvmtiAddCapabilities()` otherwise they will be disabled by JIT, for example, `can_generate_method_exit_events/JVMTI_EVENT_METHOD_EXIT` related hooks are `J9HOOK_VM_METHOD_RETURN/J9HOOK_VM_NATIVE_METHOD_RETURN` https://github.com/eclipse-openj9/openj9/blob/c24c9542ffd415e5431ae2a2109129f6c55e4268/runtime/jvmti/jvmtiHook.c#L692-L694
The hook `J9HOOK_VM_METHOD_RETURN` is to be disabled if not reserved in advance
https://github.com/eclipse-openj9/openj9/blob/c24c9542ffd415e5431ae2a2109129f6c55e4268/runtime/compiler/control/J9Options.cpp#L3169


Similarly for the other following capabilities, if they are not reserved and a `jdwp` agent is specified at the restore option file, the capability in question can't be added and the debugger won't be able to run.

Reply:
okay thanks

Please add some of that detail in the comment

Reply:
Yup, added a comment.